File: src_kt/Crypto/SodiumStreamCrypto.kt
import java.util.Arrays // For copyOfRange if needed, though direct array manipulation is more likely for JNI style

// Assuming Libsodium.kt and StreamCrypto.kt (placeholder) are available.

// --- Placeholder for StreamCrypto interface ---
// This should be in its own file: StreamCrypto.kt (or StreamCryptoProtocol.kt)
interface StreamCrypto {
    fun update(data: ByteArray): ByteArray // Original Swift modified in-place
}
// --- End Placeholder ---


class SodiumStreamCrypto(
    val key: ByteArray,
    val iv: ByteArray, // Libsodium often calls this 'nonce'
    val algorithm: Algorithm
) : StreamCrypto {

    enum class Algorithm {
        CHACHA20, SALSA20
    }

    private var currentCounter: ULong = 0uL // Corresponds to 'ic' (initial counter) in libsodium, in blocks.
                                          // The Swift 'counter' was byte-based.
                                          // crypto_stream_..._xor_ic takes block counter.
    private val blockSize: Int = 64 // Bytes, as per crypto_stream_* block size

    init {
        // Ensure Libsodium is initialized (conceptual call)
        Libsodium.ensureInitialized()

        // TODO: Validate key and IV/nonce lengths based on the algorithm
        // e.g., ChaCha20 key is 32 bytes. Nonce is typically 8, 12 (standard RFC8439), or 24 (XChaCha20).
        // Salsa20 key is 32 bytes. Nonce is typically 8 or 24.
        // The passed 'iv' here is used as nonce.
    }

    /**
     * Encrypts or decrypts data using the stream cipher.
     *
     * TODO: The padding logic from the Swift code is unusual for stream ciphers and needs review.
     * Stream ciphers like ChaCha20/Salsa20 operate as XOR streams and do not inherently require
     * input to be padded to block boundaries. The Swift code's padding might be related to how
     * it manages the counter or specific requirements of its usage context.
     * This Kotlin version attempts to replicate the logic but it might be simplified if
     * standard stream cipher application is intended.
     *
     * The original Swift code modified data in-place. This version returns a new ByteArray.
     */
    override fun update(data: ByteArray): ByteArray {
        if (data.isEmpty()) {
            return ByteArray(0)
        }

        // Calculate current byte offset within the conceptual keystream block
        val bytePaddingInBlock = (currentCounter * blockSize.toULong()) % blockSize.toULong() // This is byte offset from start of stream
                                                                                          // The swift `counter % blockSize` refers to total bytes processed % blocksize

        // The Swift code's "padding" seems to be about aligning the *input data* to the current position
        // in the keystream block if previous operations left `counter` not block-aligned.
        // This is complex. Let's simplify: `crypto_stream_xor_ic` takes a block counter `ic`.
        // We need to manage the total bytes processed (`totalBytesProcessed`) to correctly derive `ic` and
        // the offset *within* the current keystream block for the current `data` chunk.

        // For this translation, we'll assume a direct call to a hypothetical JNI/wrapper.
        // The output buffer for XORing.
        val outputData = data.copyOf() // Work on a copy to return a new array

        // Calculate initial block counter for this specific call
        // `currentCounter` here is the total bytes processed so far by this stream instance.
        val initialBlockForThisCall = currentCounter / blockSize.toULong()
        val offsetInKeystreamBlock = (currentCounter % blockSize.toULong()).toInt()


        // TODO: This is where the actual call to Libsodium (via JNI/JNA) or a JVM equivalent happens.
        // The following is a conceptual representation.
        // A real implementation needs a robust Libsodium binding or a JCE cipher.
        // For JCE, ChaCha20 might be available (e.g. "ChaCha20" or "ChaCha20-Poly1305" in newer JDKs,
        // or from BouncyCastle "ChaCha7539"). Salsa20 is usually from BouncyCastle.
        // These ciphers would need to be initialized in encrypt/decrypt mode with key, nonce, and initial counter.
        // Managing the stream and counter for partial updates with JCE Cipher can be tricky.

        val success = when (algorithm) {
            Algorithm.CHACHA20 -> {
                // Hypothetical JNI call:
                // nativeCryptoStreamChaCha20XorIc(outputData, 0, outputData.size, /*inputData*/ outputData, 0, outputData.size, iv, initialBlockForThisCall, key)
                System.err.println("TODO: Call crypto_stream_chacha20_xor_ic via JNI/JNA or use JCE ChaCha20")
                // Example using a hypothetical JCE setup (would need proper init and counter handling):
                // val cipher = Cipher.getInstance("ChaCha20") // Potentially BouncyCastle
                // val params = ChaCha20ParameterSpec(iv, initialBlockForThisCall.toInt()) // Nonce, counter
                // cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "ChaCha20"), params)
                // cipher.update(data, 0, data.size, outputData, 0)
                true // Assume success for placeholder
            }
            Algorithm.SALSA20 -> {
                // Hypothetical JNI call:
                // nativeCryptoStreamSalsa20XorIc(outputData, 0, outputData.size, /*inputData*/ outputData, 0, outputData.size, iv, initialBlockForThisCall, key)
                System.err.println("TODO: Call crypto_stream_salsa20_xor_ic via JNI/JNA or use JCE Salsa20")
                true // Assume success for placeholder
            }
        }

        if (!success) {
            throw RuntimeException("Cryptographic operation failed for $algorithm")
        }

        currentCounter += data.size.toULong() // Update total bytes processed

        // The Swift code's padding copy-back logic is omitted here because if the crypto
        // operation is done correctly (XORing input with keystream), the outputData
        // is the result. The complex padding in Swift seemed to be compensating for something
        // in how it was feeding data or managing the counter with `crypto_stream_*_xor_ic`.
        // A direct application of stream cipher should not need that kind of input padding/output unpadding.
        // If the `_ic` function requires data to be block-aligned *relative to the start of the stream*,
        // that's a different constraint that needs careful handling of partial data blocks.
        // However, `crypto_stream_xor_ic` is designed to handle arbitrary lengths and offsets via its counter `ic`.

        return outputData
    }
}

File: swift/src/Crypto/SodiumStreamCrypto.swift
import Foundation
#if canImport(Clibsodium)
import Clibsodium
#endif
import Sodium

open class SodiumStreamCrypto: StreamCryptoProtocol {
    public enum Alogrithm {
        case chacha20, salsa20
    }

    public let key: Data
    public let iv: Data
    public let algorithm: Alogrithm

    var counter = 0

    let blockSize = 64

    public init(key: Data, iv: Data, algorithm: Alogrithm) {
        _ = Libsodium.initialized
        self.key = key
        self.iv = iv
        self.algorithm = algorithm
    }

    open func update(_ data: inout Data) {
        guard data.count > 0 else {
            return
        }

        let padding = counter % blockSize

        var outputData: Data
        if padding == 0 {
            outputData = data
        } else {
            outputData = Data(count: data.count + padding)
            outputData.replaceSubrange(padding..<padding + data.count, with: data)
        }

        var tempOutput = outputData
        let count = tempOutput.count
        switch algorithm {
        case .chacha20:
            _ = tempOutput.withUnsafeMutableBytes { outPtr in
                _ = outputData.withUnsafeBytes { inPtr in
                    iv.withUnsafeBytes { ivPtr in
                        key.withUnsafeBytes { keyPtr in
                            crypto_stream_chacha20_xor_ic(outPtr.baseAddress!, inPtr.baseAddress!, UInt64(count), ivPtr.baseAddress!, UInt64(counter/blockSize), keyPtr.baseAddress!)
                        }
                    }
                }
            }

        case .salsa20:
            _ = tempOutput.withUnsafeMutableBytes { outPtr in
                _ = outputData.withUnsafeBytes { inPtr in
                    iv.withUnsafeBytes { ivPtr in
                        key.withUnsafeBytes { keyPtr in
                            crypto_stream_salsa20_xor_ic(outPtr.baseAddress!, inPtr.baseAddress!, UInt64(count), ivPtr.baseAddress!, UInt64(counter/blockSize), keyPtr.baseAddress!)
                        }
                    }
                }
            }
        }
        outputData = tempOutput

        counter += data.count

        if padding == 0 {
            data = outputData
        } else {
            data.withUnsafeMutableBytes {
                outputData.copyBytes(to: $0, from: padding..<outputData.count)
            }
        }
    }
}
